shared/src/commonMain/kotlin/org/hyperskill/app/step_completion/presentation/StepCompletionActionDispatcher.kt

```diff
@@ -34,14 +37,19 @@ class StepCompletionActionDispatcher(
     private val sentryInteractor: SentryInteractor,
     private val freemiumInteractor: FreemiumInteractor,
     private val nextLearningActivityStateRepository: NextLearningActivityStateRepository,
+    private val currentProfileStateRepository: CurrentProfileStateRepository,
     private val topicCompletedFlow: TopicCompletedFlow,
     private val topicProgressFlow: TopicProgressFlow,
-    notificationInteractor: NotificationInteractor,
+    private val notificationInteractor: NotificationInteractor,
 ) : CoroutineActionDispatcher<Action, Message>(config.createConfig()) {
     init {
         notificationInteractor.solvedStepsSharedFlow
-            .onEach { stepId ->
-                onNewMessage(Message.StepSolved(stepId))
+            .onEach { solvedStepId ->
+                if (notificationInteractor.isRequiredToAskUserToEnableDailyReminders()) {
+                    onNewMessage(Message.RequestDailyStudyRemindersPermission)
+                } else {
+                    checkProblemOfDaySolved(solvedStepId)
+                }
```

Comment:
We have a very old bug:
1. User solved a problem of the day
2. We requested daily study reminders permission
```
if (notificationInteractor.isRequiredToAskUserToEnableDailyReminders()) {
    onNewMessage(Message.RequestDailyStudyRemindersPermission)
}
```
3. After solving the problem of the day a user receives some gems, but we did not update its hypercoins balance, so by navigating back a user will see the old hypercoins balance until content will be updated at some point.

So, if the user solved a problem of the day we always need to refresh its balance to hold it up to date.
Problem of the day solved bottom sheet presentation should be performed only when we are not requested daily study reminders permission.

Reply:
@ivan-magda by using `currentProfileStateRepository` in this action dispatcher we solved problem with updating gems. Code you highlighted is guaranteeing that problem of day solved bottom sheet presentation is performed only when we are not requested aily study reminders permission.

Reply:
@vladkash there is a bug, try to understand https://github.com/hyperskill/mobile-app/pull/610#discussion_r1289517900

we are not updating hypercoins balance when problem of the day solved -> permission alert presented